Add the start of AbsolutePath and RelativePath types for path handling

abertelrud opened the pull request:
This will provide a more type safe and expressive path API than plain
strings.  Many more details are in the documentation of the types.

More extensive correctness and performance tests, as well as adoption,
to come.  There is still a fair amount of performance investigation and
tuning to do -- that can be addressed piecemeal, but the direction of
the API (in particular the distinction between absolute and relative
paths) is the most important part at this point.

abertelrud commented:
One comment about the names here:  since most paths should be absolute, it is very probably that `AbsolutePath` will be renamed to just `Path` once I've finished switching over the source base to use it, and removed the previous `Path` class in Utility.  Partly the `AbsolutePath` name right now is to prevent confusion.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+
+/// Path separator character (always `/`).  If a path separator occurs at the
+/// beginning of a path, the path is considered to be absolute.
+public let PathSeparatorCharacter: Character = "/"
```
Do these need to be public, or can they be private? Also, I think they should probably be initial-downcased since they are values not types.

abertelrud replied:
They can be private.  I didn't realize that globals should be lowercase, but I'll make them that.  I'll make the change.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+        var chars = string.characters
+        
+        // Expand any `~` or `~user` prefix.
+        if chars.first == HomeDirectoryCharacter {
```
I would prefer if this method didn't do home-path resolution, I don't think it is the best default for a core type, since it is (a) inconsistent with most UNIX tools, and (b) is not purely syntactic, since in the general case it may need to communicate with the network.

abertelrud replied:
I see your point in theory, but that means that every client needs to remember to do this resolution individually on any path that can come in from the user.  That could lead to easily lead to cases where we end up with unexpanded paths starting with `~` characters.

Perhaps there is a pragmatic distinction to be made between ~ and ~user, since the former is not likely to require network access (even if the user's home directory is on the network, the actual characters that comprise the path are available locally; ~user is of course different, and perhaps that should be disallowed).

ddunbar replied:
That is true, but for a command line tool usually `~` expansion would be performed on sanitization of a command line provided path. We already are going to have to have a routine to perform that sanitization since all command line relative paths need to become `AbsolutePath`s, so that would be once place to concentrate this normalization.

abertelrud replied:
That's a good point — it makes sense to separate out the API that takes a "string that might represent an absolute or relative path, possible with a `~` prefix, and makes it into an AbsolutePath value" from the API that takes a "string that's expected to represent an absolute path and makes it into an AbsolutePath value".

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+        // Get a hold of the character view.
+        // FIXME: We need to investigate what is most performant here.
+        //        For example, is UTF-8 view more efficient, or unichars, etc?
+        var chars = string.characters
```
We should use UTF8.

abertelrud replied:
Is it more efficient?

abertelrud replied:
Agreed.  I wanted to start with the most obvious view and then measure to see if we needed to go to UTF-8, but since we're dealing with strings in which all the interesting characters are ASCII, going directly to UTF-8 makes sense.  I'll swift it over.

abertelrud replied:
One significant drawback of String.UTF8View is that it isn't bidirectional, so it's very inefficient to do things like find the last occurrence of, say, a path separator.  That's a problem.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+        // the normalized string representation.
+        var parts: [String.CharacterView] = []
+        var capacity = 0
+        for part in chars.split(separator: PathSeparatorCharacter) {
```
Splitting on each component, building an array of parts, and reassembling is going to be somewhat expensive. It would be nice if we had a fast path here for cases where the input string is already normalized, can we get a FIXME for that (or the fast path, if it is easy)?

abertelrud replied:
Yes, there is a lot of optimization that can be done here, and the idea is to guide that by profiling the performance tests.  I agree that we're very likely to want to optimize for this case.  I'll add a FIXME.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+            "\(result.characters.count), cap: \(capacity)")
+        
+        // Use the result as our stored string.
+        _impl = PathImpl(string: String(result))
```
`String(result)` is unnecessary, right?

abertelrud replied:
Probably.  I'll check.  `result` used to not be a string, and therefore it was necessary.  But it probably is no longer.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+        return _impl.dirname!
+    }
+    
+    /// Last path component (including the suffix, if any).  it is never empty.
```
It is empty for the base name of the root path, right?

abertelrud replied:
I didn't think it was, but I suppose this is a matter of debate.

ddunbar replied:
Oh, I see -- I hadn't read the implementation yet. I personally would expect the base name to never contain "/" but I guess that was just my assumption. I don't have a strong opinion on the right answer here, I see that the POSIX utility says it is "/" while most APIs I have used say "".

abertelrud replied:
Yes, I was looking at things like `dirname(1)` and `basename(1)` for semantics, and also Foundation's own `pathComponents` method, which returns `/` as the first path component of an absolute path (implying that that is its "name").  The root directory will always be kind of weird in this way, since it's also possible to have an arbitrary number of `..` path components that still just resolve to itself.  Also Ruby's Pathname (for example) returns `/` as both the base name and the directory name of `/`.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+    /// Root directory (whose string representation is just a path separator).
+    public static var root: AbsolutePath {
+        // FIXME: If this is called a lot, we should maybe cache the root path.
+        return AbsolutePath(String(PathSeparatorCharacter))
```
Why not go ahead and do this? It seems free.

abertelrud replied:
It isn't free, because of thread safety (or can this be created statically, because it's a value type? — if so then I agree we should just do it, though I don't know how often this will get called).

ddunbar replied:
It can be assigned statically (the compiler *may* use `dispatch_once` under the covers).

abertelrud replied:
Cool.  I'll make the change.  I'd be surprised if the compiler would need `dispatch_once`, since it seems that the bits could be laid out in memory (assuming that short strings are inlined, which, come to think of it, I haven't checked if they are in Swift).  Either way, that seems like an orthogonal optimization that would happen under-the-covers.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+    /// Returns the absolute path with the relative path applied.
+    public func join(_ subpath: RelativePath) -> AbsolutePath {
+        let slash = String(PathSeparatorCharacter)
+        return AbsolutePath(asString + slash + subpath.asString)
```
It would be good to have a fast path here too, otherwise the default behavior for joining a sequence of paths is O(N^2).

abertelrud replied:
Yes, there will definitely be a fast path here.  There is a comment about that right below.

ddunbar replied:
What I was specifically referring to was a fast path over `asString + slash + subpath.asString` to not re-normalize the starting path... the comment below seemed to be about something else.

abertelrud replied:
Ah, now I see what you mean.  Yes, that's a clear oversight.  I'll fix that.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+                capacity += part.count
+            }
+        }
+        capacity += max(parts.count - 1, 0)
```
Can we eliminate the code duplication with the `AbsolutePath` initializer?

abertelrud replied:
This isn't really code duplication except for a few lines.  I started off writing a single method that handled both absolute and relative paths, but because of differences in the semantics of `..` at the beginning of absolute vs relative paths, the unified code turned out to be a lot more complicated.  So I split it out.

ddunbar replied:
Ok, makes sense.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+    }
+    
+    /// Suffix (including leading `.` character) if any.  Note that a basename
+    /// that starts with a `.` character is not considered a suffix.
```
This comment isn't correct, the implementation doesn't include the `.`.

abertelrud replied:
Correct.  I forgot to address the `FIXME: This needs to be rewritten` (this code came from the old Path implementation).  I'll fix it.

abertelrud replied:
In the new code, this comment is now correct.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+        let parts = chars.split(separator: ".")
+        if let last = parts.last where parts.count > 1 { return String(last) }
+        return nil
+    }
```
This shouldn't return a suffix for "/a/b.c/d", right?

abertelrud replied:
Right.  I forgot to address the `FIXME: This needs to be rewritten` (this code came from the old Path implementation).  I'll fix it.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -0,0 +1,473 @@
+    public var isFile: Bool {
+        var mystat = stat()
+        let rv = lstat(self.asString, &mystat)
+        return rv == 0 && (mystat.st_mode & S_IFMT) == S_IFREG
```
Do we want to use `lstat` in these two methods?

abertelrud replied:
Yes.  Or rather, this whole API needs to be redone, as noted in the FIXME.  The user needs to be able to choose whether or not to examine the entity at the path vs any resolved symlink.  The old code which always used stat() was definitely wrong, since it could never return true for isSymlink.

abertelrud replied:
I had wanted to pend reworking this until after landing the basic path support, but I will either do it here or pull this API for now.

ddunbar replied:
The old code used `lstat` for `isSymlink`, but `stat` for `isDirectory` and `isFile`.

ddunbar commented:
One last comment aside from the inline ones: I'm a little unsure if it is worth using optionals for `dirname` and `basename` (versus empty strings). In particular, having `AbsolutePath` and `RelativePath` have different result types for the same named APIs might be annoying if we want a unified protocol over them.

abertelrud commented:
I agree.  I was on the fence about having it be optional or return an empty string.  In any other language it would have been an empty string, but in this case it felt natural to say it has "no" parent directory.  But I agree with your point about a potential future protocol, so I'll change them to be non-optional.

abertelrud commented:
Thanks for all your comments — I'll update the PR with a revised set of diffs as soon as they're ready.

abertelrud commented:
I have updated the PR with the latest version.  To reduce the change to just the most basic thing we need, and avoid some of the issues with home path resolution, I have removed that support for now.  I have also (temporarily!) removed the methods to access the file system, so that this source file deals purely with the data structure aspects.  There will be an extension on `AbsolutePath` to add shorthand for accessing things like `stat` information.

There are FIXME comments in the source about additional methods to add.  I will add them as we have a need for them, since the important thing is to quickly get onto using these structs throughout the codebase.

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -14,128 +14,59 @@ import POSIX
+        let absStr = absPath._impl.string
+        let relStr = relPath._impl.string
+        _impl = PathImpl(string: relStr == "." ? absStr :
+            absStr + String(pathSeparatorCharacter) + relStr)
```
This needs to also handle the case where the relative path starts with "..", I think?

abertelrud replied:
Indeed.  I had actually noticed this when adding more unit tests, and am preparing an update.  Thanks!

ddunbar commented on Sources/Basic/Path.swift:
```diff
@@ -342,132 +209,244 @@ private struct PathImpl {
+        // Find the last path separator.
+        guard let idx = chars.rindex(of: pathSeparatorCharacter) else {
+            // No path separators, so the directory name is empty.
+            return ""
```
Would it make sense for dirname to return `"."` here? That seems like it might be consistent with the handling of `RelativePath("")`.

abertelrud replied:
Yes definitely, and also, that's what dirname(1) does.  I'll fix it.

ddunbar commented:
That sounds like a good plan to me... PR looks reasonable to me in this form (with the ".." fix).

abertelrud commented:
Please let me know when it's ok to merge.

ddunbar commented:
LGTM

ddunbar commented:
@swift-ci please test

abertelrud commented:
Looks like all the tests have passed.  Unless there are objections, I will merge this in.  I'll leave it open for a little in case there are any further comments, since this has been an active PR.  BTW, thanks Daniel for all the feedback.

aciidb0mb3r commented on Tests/Basic/PathTests.swift:
```diff
@@ -0,0 +1,177 @@
+    
+    // FIXME: We also need test for stat() operations.
+        
+    static var allTests = [
```
this needs to be added to Basic/XCTestManifests.swift to be able to run on Linux

abertelrud replied:
Thanks.  The CIs passed, but I guess that's because they didn't end up running these tests?

aciidb0mb3r replied:
Yea they must have been skipped for Linux

abertelrud replied:
I've checked in a fix.
